modules/standard/IO.chpl

```diff
@@ -3977,125 +3992,209 @@ proc channel.readline(ref arg: ?t): bool throws where t==string || t==bytes {
-            this._revert();
-            try this._ch_ioerror(EFORMAT:syserr, "line longer than maxSize in channel.readLine(ref s: string)");
-          }
-        }
+  on this.home {
+    try this.lock(); defer { this.unlock(); }
+    param newLineChar = 0x0A; // ascii newline.
+    var maxCodepoints = if maxSize < 0 then max(int) else maxSize;
+    var nCodepoints: int = 0; // num codepoints, including newline
+    var chr : int(32);
+    var err: syserr = ENOERR;
+    var foundNewline = false;
+    // use the channel's buffering to compute how many bytes/codepoints
+    // we are reading
+    this._mark();
+    while !foundNewline {
+      // read a single codepoint
+      err = qio_channel_read_char(false, this._channel_internal, chr);
+      if err == EEOF {
+        // encountered EOF
+        break;
+      } else if err {
+        // encountered an error so throw
         this._revert();
+        try this._ch_ioerror(err, "in channel.readLine(ref s: string)");
       }
-      var saveStyle: iostyleInternal = this._styleInternal();
-      defer {
-        this._set_styleInternal(saveStyle);
+      nCodepoints += 1;
+      if chr == newLineChar {
+        foundNewline = true;
       }
-      var myStyle = saveStyle.text();
-      myStyle.string_format = QIO_STRING_FORMAT_TOEND;
-      myStyle.string_end = newLineChar;
-      if(maxSize >= 0){
-        try myStyle.max_width_characters = maxSize.safeCast(uint(32)); // Probably should change maxSize's type to be safe to cast
+      if nCodepoints > maxCodepoints {
+        if stripNewline && foundNewline {
+          // don't worry about it since we wouldn't return the newline
+        } else {
+          // The line is longer than was specified so we throw an error
+          this._revert();
+          try this._ch_ioerror(EFORMAT:syserr,
+               "line longer than maxSize in channel.readLine(ref s: string)");
+        }
       }
-      this._set_styleInternal(myStyle);
-      try _readOne(iokind.dynamic, s, origLocale);
     }
-  } catch err: SystemError {
-    if err.err != EEOF then throw err;
-    return false;
-  }
-  // TODO find a more memory efficient way to do this
-  // Maybe have a QIO_STRING_FORMAT_TOEND_DROPEND which drops the last char
-  if(stripNewline){
-    s = s.strip(chars="\n", leading = false);
+    var endOffset = this._offset();
+    this._revert();
+    var nBytes:int = endOffset - this._offset();
+    // now, nCodepoints and nBytes include the newline
+    if foundNewline && stripNewline {
+      // but we don't want to read the newline if stripNewline=true.
+      nBytes -= 1;
+      nCodepoints -= 1;
+    }
+
+    // now read the data into the string
+    err = readStringBytesData(s, this._channel_internal, nBytes, nCodepoints);
+    if foundNewline && stripNewline && !err {
+      // pass the newline in the input
+      err = qio_channel_read_char(false, this._channel_internal, chr);
+    }
```

Comment:
It took me a while to grok what's happening here. Is there way we can avoid the `_revert` and re-reading the codepoint we already have read?

But that maybe a bigger change than you'd like here. If that's the case can we add a comment right before calling `readStringBytesData` saying something like

```
readStringBytesData will advance the channel by exactly `nBytes`.
This may consume or leave the newline based on the logic above.
```

Reply:
I'll add the suggested comment. I don't know (right now) of a way to avoid the revert and re-read since we are computing the string length that way.